### PR TITLE
Fix dominant light shadows in Metal and Vulkan

### LINKED ISSUE #3619: Dominant Light Shadows on Metal
**Describe the bug**
Dominant Light Shadows feature looks different and glitchy on Metal compared to OpenGL.

**To Reproduce**
Compare the same models with the same lighting parameters on OpenGL and Metal in the gltf_viewer sample.

**Expected behavior**
Dominant Light Shadows should look the same on Metal and OpenGL.

**Screenshots**
See added screenshot.

**Desktop (please complete the following information):**
 - OS: macOS Big Sur 11.2.1
 - GPU: Intel Iris Plus Graphics 645
 - Backend: OpenGL/Metal

**Additional context**
On Metal seems like the direction of DLS is a different vector and the camera angle has a much more dramatic influence on it than on OpenGL.

dominant-shadows
.

### PR DESCRIPTION
**Pull request by bejado**
SAO calculations rely on two coordinate spaces:
- UV-space: framebuffer/texture space, graphics API dependent.
- screen-space: origin bottom left, normalized coordinates.

We didn't see any issues with OpenGL, as its UV-space matches screen-space (Y up). However, Metal and Vulkan's UV-space is vertically flipped (Y down).

This is my attempt to resolve the discrepancy by adding in conversions between the spaces when necessary. @pixelflinger I'm curious if you have other ideas or prefer different terminology.

Fixes #3619

### CODE REVIEW

File: filament/src/materials/ssao/sao.mat [hunk 1]
```diff
@@ -153,8 +153,16 @@ fragment {
         return fract(m.z * fract(dot(w, m.xy)));
     }
 
-    highp vec3 computeViewSpacePositionFromDepth(highp vec2 uv, highp float linearDepth) {
-        return vec3((0.5 - uv) * materialParams.positionParams.xy * linearDepth, linearDepth);
+    highp vec2 getApiAgnosticFragCoords() {
```
Comment (romainguy) [inline filament/src/materials/ssao/sao.mat:156]: Maybe just getFragCoord()? (at least drop the s at the end)

File: filament/src/materials/ssao/ssaoUtils.fs [hunk 2]
```diff
@@ -13,5 +13,16 @@ highp float sampleDepthLinear(const sampler2D depthTexture, const vec2 uv, float
     return linearizeDepth(textureLod(depthTexture, uv, lod).r, depthParams);
 }
 
+highp vec2 ssFromUv(highp vec2 uv) {
```
Comment (romainguy) [inline filament/src/materials/ssao/ssaoUtils.fs:16]: I read this as "screenSpace", it's not really screen space, it's still a UV. Not sure what a better name would be. screenUvFromUv()?

File: filament/src/materials/ssao/sao.mat (hunk 1 above)
Comment (pixelflinger) [reply filament/src/materials/ssao/sao.mat:156]: +1 for  `getFragCoord`

File: filament/src/materials/ssao/sao.mat [hunk 3]
```diff
@@ -316,7 +332,7 @@ fragment {
 
 #if defined(TARGET_MOBILE)
         // this line is needed to workaround what seems to be a bug on qualcomm hardware
-        aoVisibility += gl_FragCoord.x * MEDIUMP_FLT_MIN;
+        aoVisibility += getApiAgnosticFragCoords().x * MEDIUMP_FLT_MIN;
```
Comment (pixelflinger) [inline filament/src/materials/ssao/sao.mat:335]: this one is just a workaround for qualcomm, I would just leave it.

File: filament/src/materials/ssao/ssaoUtils.fs (hunk 2 above)
Comment (pixelflinger) [reply filament/src/materials/ssao/ssaoUtils.fs:16]: Agreed, I also thought it was "screen-space" -- but it's like a normalized screenspace, isn't that "NDC", Normalized device coordinate?
This method needs a comment about what the input/output are.
Maybe `NdcFromUv` ?

File: filament/src/materials/ssao/ssaoUtils.fs [hunk 4]
```diff
@@ -13,5 +13,16 @@ highp float sampleDepthLinear(const sampler2D depthTexture, const vec2 uv, float
...
+#if defined(TARGET_METAL_ENVIRONMENT) || defined(TARGET_VULKAN_ENVIRONMENT)
+    return vec2(uv.x, 1.0 - uv.y);
+#endif
+    return uv;
+}
+
+highp vec2 uvFromSs(highp vec2 ss) {
+    return ssFromUv(ss);
```
Comment (pixelflinger) [inline filament/src/materials/ssao/ssaoUtils.fs:24]: I would just reimplement the code here -- it's super confusing to call the opposite function.

File: filament/src/materials/ssao/sao.mat (hunk 1 above)
Comment (pixelflinger) [reply filament/src/materials/ssao/sao.mat:156]: We need a comment that explains that it's returning in the gl convention (0,0) at bottom-left.

File: filament/src/materials/ssao/sao.mat [hunk 5]
```diff
@@ -287,18 +298,23 @@ fragment {
...
 
+    // For SAO calculations, we make use of two coordinate spaces:
+    //     UV-space: framebuffer/texture space, graphics API dependent.
+    //     screen-space: origin bottom left, normalized coordinates.
+    // Textures must always be sampled with UV coordinates.
+
     void postProcess(inout PostProcessInputs postProcess) {
         highp vec2 uv = variable_vertex.xy; // interpolated to pixel center
 
         highp float depth = sampleDepthLinear(materialParams_depth, uv, 0.0, materialParams.depthParams);
```
Comment (pixelflinger) [inline filament/src/materials/ssao/sao.mat:316]: if you made `sampleDepthLinear` take a `ss` you'd be able to remove a lot of calls to `uvFromSs` -- might simplify things.

Comment (pixelflinger) [reply filament/src/materials/ssao/sao.mat:316]: Same thing with `computeViewSpaceNormal`.

I feel like we should do all the work in `ss` and pass `ss` everywhere, except at texture sampling time, where you can do a `uvFromSs`.

**Review by pixelflinger — CHANGES_REQUESTED** [review]

File: filament/src/materials/ssao/ssaoUtils.fs (hunk 2 above)
Comment (romainguy) [reply filament/src/materials/ssao/ssaoUtils.fs:16]: It's not NDC either; NDC uses -1…1 coordinates, here it's 0…1.

Comment (bejado) [reply filament/src/materials/ssao/ssaoUtils.fs:16]: Yeah this isn't NDC. It's a left hand, normalized screen space. I'm fine with using "screenUV" as the terminology, but I'm not sure how to abbreviate it. `suv`?

**Comment by bejado** [comment]
So, I ended up making this more complicated than necessary. There's no need to introduce any new terminology. On Metal/Vulkan, we can simply flip the UV coordinates in the vertex shader and then flip again when we go to sample the depth texture.

**Review by pixelflinger — APPROVED** [review]
much nicer! Thanks!

**Pull request merged**